src/util/ringdelaybuffer.h

```diff
@@ -0,0 +1,147 @@
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR
+* ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
+* CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+* WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+/*
+* The text above constitutes the entire PortAudio license; however,
+* the PortAudio community also makes the following non-binding requests:
+*
+* Any person wishing to distribute modifications to the Software is
+* requested to send the modifications to the original developer so that
+* they can be incorporated into the canonical version. It is also
+* requested that these non-binding requests be included along with the
+* license above.
+*/
+
+#pragma once
+
+// The RingDelayBuffer code is based on the pa_ringbuffer.c
+// from the Portable Audio I/O Library. The template code is used
+// in the Mixxx.
+//
+// For the pa_ringbuffer.c see: mixxx/lib/portaudio/pa_ringbuffer.c
+//
+// Added functions:
+// - isFull
+// - isEmpty
+// - length
+// - getReadAvailable
+// - getWriteAvailable
+// - moveReadPositionBy
+//
+// Modified functions:
+// - clear
+// - read
+// - write
+
+#include "util/samplebuffer.h"
+#include "util/types.h"
+
+// The RingDelayBuffer is designed to be single-threaded
+// and based on that IS NOT THREAD SAVE.
+
+// TODO(davidchocholaty) Done documentation.
+
+/// The ring delay buffer allows moving with the reading position subject
+/// to certain rules. Another difference between the classic ring buffer is,
+/// that the ring delay buffer offers to read zero values
+/// which were not written by using the write method and write position.
+/// Both of these two specific properties are based on the cross-fading
+/// between changes of two delays.
+class RingDelayBuffer final {
+  public:
+    RingDelayBuffer(SINT bufferSize);
+
+    bool isFull() const {
+        return getWriteAvailable() == 0;
+    }
+
+    bool isEmpty() const {
+        return getReadAvailable() == 0;
+    }
+
+    void clear() {
+        m_fullFlag = false;
+        m_firstInputBuffer = true;
+
+        m_readPos = 0;
+        m_writePos = 0;
+
+        m_buffer.fill(0);
+    }
+
+    SINT size() const {
+        return m_buffer.size();
+    }
+
+    SINT getReadAvailable() const {
+        if (m_writePos > m_readPos) {
+            return (m_writePos - m_readPos);
+        } else if (m_writePos < m_readPos) {
+            return (m_buffer.size() - m_readPos) + m_writePos;
+        } else {
+            // The write position equals the read position (m_writePos == m_readPos).
+            // The buffer is full or empty (m_fullFlag).
+            if (m_fullFlag) {
+                return m_buffer.size();
+            }
+
+            return 0;
+        }
+    }
```

Comment:
I came up with this optimization to minimize branch misspredictions. Idk if it actually improves things. I guess just try locally and test if you see a difference in the benchmarks.
```suggestion
    SINT getReadAvailable() const {
        const SINT delta = m_writePos - m_readPos;
        if (delta == 0) [[unlikely]] {
          return m_fullFlag ? m_buffer.size() : 0;
        }
        const SINT range = std::abs(delta);
        return range == delta ? range : m_buffer.size() - range;
    }
```

Reply:
If you are interested in it, I have the results on my system:

**Previous version**
```
----------------------------------------------------------------------------------------------
Benchmark                                                    Time             CPU   Iterations
----------------------------------------------------------------------------------------------
BM_WriteReadWholeBufferNoSkip/64_mean                      528 ns          531 ns           20
BM_WriteReadWholeBufferNoSkip/64_median                    528 ns          531 ns           20
BM_WriteReadWholeBufferNoSkip/64_stddev                  0.513 ns        0.478 ns           20
BM_WriteReadWholeBufferNoSkip/64_cv                       0.10 %          0.09 %            20
BM_WriteReadWholeBufferNoSkip/512_mean                     612 ns          616 ns           20
BM_WriteReadWholeBufferNoSkip/512_median                   613 ns          616 ns           20
BM_WriteReadWholeBufferNoSkip/512_stddev                 0.880 ns        0.988 ns           20
BM_WriteReadWholeBufferNoSkip/512_cv                      0.14 %          0.16 %            20
BM_WriteReadWholeBufferNoSkip/4096_mean                   1339 ns         1370 ns           20
BM_WriteReadWholeBufferNoSkip/4096_median                 1341 ns         1373 ns           20
BM_WriteReadWholeBufferNoSkip/4096_stddev                 6.57 ns         8.56 ns           20
BM_WriteReadWholeBufferNoSkip/4096_cv                     0.49 %          0.62 %            20
BM_WriteReadWholeBufferSkipLeftNoCircle/64_mean            535 ns          539 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/64_median          534 ns          538 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/64_stddev         1.03 ns        0.994 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/64_cv             0.19 %          0.18 %            20
BM_WriteReadWholeBufferSkipLeftNoCircle/512_mean           617 ns          621 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/512_median         616 ns          621 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/512_stddev        2.98 ns         2.97 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/512_cv            0.48 %          0.48 %            20
BM_WriteReadWholeBufferSkipLeftNoCircle/4096_mean         1371 ns         1397 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/4096_median       1375 ns         1405 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/4096_stddev       9.64 ns         12.8 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/4096_cv           0.70 %          0.91 %            20
BM_WriteReadWholeBufferSkipLeftCircle/64_mean              538 ns          542 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/64_median            538 ns          541 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/64_stddev           1.37 ns         1.20 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/64_cv               0.25 %          0.22 %            20
BM_WriteReadWholeBufferSkipLeftCircle/512_mean             621 ns          626 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/512_median           621 ns          626 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/512_stddev         0.396 ns        0.412 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/512_cv              0.06 %          0.07 %            20
BM_WriteReadWholeBufferSkipLeftCircle/4096_mean           1427 ns         1467 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/4096_median         1456 ns         1497 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/4096_stddev         40.4 ns         40.9 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/4096_cv             2.83 %          2.78 %            20
```
**Optimized version**
```
----------------------------------------------------------------------------------------------
Benchmark                                                    Time             CPU   Iterations
----------------------------------------------------------------------------------------------
BM_WriteReadWholeBufferNoSkip/64_mean                      528 ns          532 ns           20
BM_WriteReadWholeBufferNoSkip/64_median                    528 ns          531 ns           20
BM_WriteReadWholeBufferNoSkip/64_stddev                   2.03 ns         1.91 ns           20
BM_WriteReadWholeBufferNoSkip/64_cv                       0.38 %          0.36 %            20
BM_WriteReadWholeBufferNoSkip/512_mean                     619 ns          623 ns           20
BM_WriteReadWholeBufferNoSkip/512_median                   618 ns          622 ns           20
BM_WriteReadWholeBufferNoSkip/512_stddev                  3.65 ns         3.61 ns           20
BM_WriteReadWholeBufferNoSkip/512_cv                      0.59 %          0.58 %            20
BM_WriteReadWholeBufferNoSkip/4096_mean                   1345 ns         1375 ns           20
BM_WriteReadWholeBufferNoSkip/4096_median                 1342 ns         1372 ns           20
BM_WriteReadWholeBufferNoSkip/4096_stddev                 7.18 ns         7.49 ns           20
BM_WriteReadWholeBufferNoSkip/4096_cv                     0.53 %          0.54 %            20
BM_WriteReadWholeBufferSkipLeftNoCircle/64_mean            532 ns          537 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/64_median          531 ns          536 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/64_stddev         2.00 ns         1.55 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/64_cv             0.38 %          0.29 %            20
BM_WriteReadWholeBufferSkipLeftNoCircle/512_mean           617 ns          621 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/512_median         616 ns          620 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/512_stddev        1.08 ns         1.07 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/512_cv            0.18 %          0.17 %            20
BM_WriteReadWholeBufferSkipLeftNoCircle/4096_mean         1366 ns         1395 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/4096_median       1363 ns         1395 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/4096_stddev       5.43 ns         3.70 ns           20
BM_WriteReadWholeBufferSkipLeftNoCircle/4096_cv           0.40 %          0.27 %            20
BM_WriteReadWholeBufferSkipLeftCircle/64_mean              537 ns          542 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/64_median            535 ns          539 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/64_stddev           5.06 ns         5.09 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/64_cv               0.94 %          0.94 %            20
BM_WriteReadWholeBufferSkipLeftCircle/512_mean             625 ns          631 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/512_median           624 ns          630 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/512_stddev          2.83 ns         2.74 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/512_cv              0.45 %          0.43 %            20
BM_WriteReadWholeBufferSkipLeftCircle/4096_mean           1387 ns         1425 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/4096_median         1388 ns         1424 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/4096_stddev         2.15 ns         2.00 ns           20
BM_WriteReadWholeBufferSkipLeftCircle/4096_cv             0.15 %          0.14 %            20
```

Reply:
Interesting. So at least its not a pessimization. What are your thoughts @daschuer?

Reply:
Sorry, but according the benchmarks, it is a pessimization. 
I think it is because we have now three branches, instead of one. The if, inside std::abs() and the conditional assignment.
Every branch is likely flushing the pipeline. The [[unlikely]] may help here though. 

The best optimization is to dispose the whole function. Because we actually have always size() readable. This is true even after initialization where we have zeros available.

Reply:
> Sorry, but according the benchmarks, it is a pessimization.

Well, IMO the difference is not clear enough to make a clear decision.

> I think it is because we have now three branches, instead of one. The if, inside std::abs() and the conditional assignment.

Optimistically speaking, the resulting code could only have a single branch.
```cpp
if (delta == 0) [[unlikely]] {
  return m_fullFlag ? m_buffer.size() : 0;
}
```
This results in a single branch which is very unlikely (the `[[unlikely]]` attribute probably doesn't change that, the branch predictor could pick that up by itself). The ternary statement afterwards is just a `cmove` (assuming `m_buffer.size()` is inlined). 
```
const SINT range = std::abs(delta);
```

Theoretically, std::abs can be implemented branchless for integers. Not sure how the standard does it.  My guess is that std::abs is probably the bottleneck if it can't get inlined (which is the case for some c apis).

```cpp
return range == delta ? range : m_buffer.size() - range;
```
this results in a branch because I was careless. If we pull out the `m_buffer.size() - range` and compute that unconditionally, this ternary could be a `cmove` as well. 


In the end, the branchless version probably has less mispredictions, though it probably suffers from data-dependency problems which also lead to bubbles in the pipeline. 

In the end, it doesn't seem to matter what method we use. If for some reason the alternative implementation wouldve been >2x faster, than I'd advocate for using that, but since it doesn't make a difference, the previous version is probably better because its code is much more readable.

Reply:
> If for some reason the alternative implementation wouldve been >2x faster, than I'd advocate for using that, but since it doesn't make a difference, the previous version is probably better because its code is much more readable.

I am on the same. It is not a problem to use the optimised version. Anyway, I expected better results for that, so, I agree to keep the first more readable version.